INT8 MKL-DNN v2 integrate to slim

zhaify opened the pull request:
Refactor the code from PR [16865](https://github.com/PaddlePaddle/Paddle/pull/16865)

please help to review
@luotao1 @wzzju @wojtuss @bingyanghuang

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/PaddlePaddle/Paddle?pullRequest=17634) <br/>All committers have signed the CLA.

zhaify commented:
@luotao1  MAC_PR_CI failed and the error message is bellow, is this the build server issue?

[02:38:51]W:	 [Step 1/1] + export PATH=/usr/local/opt/opencv@2/bin:/usr/local/bin:/Library/Frameworks/Python.framework/Versions/2.7/bin:./usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Java/JavaVirtualMachines/jdk-10.0.2.jdk/Contents/Home/bin
[02:38:51]W:	 [Step 1/1] + PATH=/usr/local/opt/opencv@2/bin:/usr/local/bin:/Library/Frameworks/Python.framework/Versions/2.7/bin:./usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Java/JavaVirtualMachines/jdk-10.0.2.jdk/Contents/Home/bin
[02:38:51]W:	 [Step 1/1] + export PATH=/usr/local/opt/opencv@2/bin:/usr/local/bin:/Library/Frameworks/Python.framework/Versions/2.7/bin:./usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Java/JavaVirtualMachines/jdk-10.0.2.jdk/Contents/Home/bin:/Users/paddle/Library/Python/2.7/bin
[02:38:51]W:	 [Step 1/1] + PATH=/usr/local/opt/opencv@2/bin:/usr/local/bin:/Library/Frameworks/Python.framework/Versions/2.7/bin:./usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Java/JavaVirtualMachines/jdk-10.0.2.jdk/Contents/Home/bin:/Users/paddle/Library/Python/2.7/bin
[02:38:51]W:	 [Step 1/1] + echo /usr/local/opt/opencv@2/bin:/usr/local/bin:/Library/Frameworks/Python.framework/Versions/2.7/bin:./usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Java/JavaVirtualMachines/jdk-10.0.2.jdk/Contents/Home/bin:/Users/paddle/Library/Python/2.7/bin
[02:38:51]W:	 [Step 1/1] + git config --global user.name 'Paddle CI_MAC'
[02:38:51]W:	 [Step 1/1] + git config --global user.email paddle_ci@example.com
[02:38:51]W:	 [Step 1/1] + set +e
[02:38:51]W:	 [Step 1/1] + git remote
[02:38:51]W:	 [Step 1/1] + grep upstream
[02:38:51]W:	 [Step 1/1] + '[' 1 == 1 ']'
[02:38:51]W:	 [Step 1/1] + git remote add upstream https://github.com/PaddlePaddle/Paddle.git
[02:38:51]W:	 [Step 1/1] + set -e
[02:38:51]W:	 [Step 1/1] + git fetch upstream
[02:38:51]W:	 [Step 1/1] fatal: unable to access 'https://github.com/PaddlePaddle/Paddle.git/': Failed to connect to 172.19.57.45 port 3128: Connection refused
[02:38:51]W:	 [Step 1/1] Process exited with code 128
[02:34:27]E:	 [Step 1/1] Process exited with code 128 (Step: Build and test (Command Line))
[02:38:51]E:	 [Step 1/1] Step Build and test (Command Line) failed
[02:38:51]i: Stopping performance monitoring process
[02:38:51]i: Performance monitoring process stopped
[02:38:51]i: Publishing performance monitoring build stages data
[02:38:51] : Publishing artifacts
[02:38:51] :	 [Publishing artifacts] Collecting files to publish: [/home/teamcity/system/perfmon/temp/104459/perfmon.csv=>.teamcity/perfmon/, /home/teamcity/temp/agentTmp/build_stages.txt=>.teamcity/perfmon/]
[02:38:51] :	 [Publishing artifacts] Publishing 2 files using [ArtifactsCachePublisher]: /home/teamcity/system/perfmon/temp/104459/perfmon.csv, /home/teamcity/temp/agentTmp/build_stages.txt => .teamcity/perfmon
[02:38:51] :	 [Publishing artifacts] Publishing 2 files using [WebPublisher]: /home/teamcity/system/perfmon/temp/104459/perfmon.csv, /home/teamcity/temp/agentTmp/build_stages.txt => .teamcity/perfmon
[02:38:51] : Publishing internal artifacts
[02:38:51] :	 [Publishing internal artifacts] Publishing 1 file using [ArtifactsCachePublisher]
[02:38:51] :	 [Publishing internal artifacts] Publishing 1 file using [WebPublisher]
[02:38:51]W: Publishing artifacts
[02:38:51] :	 [Publishing artifacts] Collecting files to publish: [output => output]
[02:38:51]W:	 [Publishing artifacts] Artifacts path 'output' not found
[02:34:27] : Build finished

luotao1 commented:
I help you rerun at first. You can rerun yourself since sometimes ci is not stable.

zhaify commented:
@luotao1 
I use the following python code to download test_data and models (url: http://paddle-inference-dist.bj.bcebos.com/int8/imagenet_val_100_tail.tar.gz) before testing TestInferQuantizeStrategy, it works fine on Linux and Windows but always failed on MAC_PR_CI with http timeout. 
Do I need to add some additional configuration to the unit test on MAC_PR_CI?
``` python
from paddle.dataset.common import download
def setUp(self):
    self.cache_folder = os.path.expanduser(
        '~/.cache/paddle/dataset/int8/download')
    if not os.path.exists(self.cache_folder):
        os.makedirs(self.cache_folder)

    self.mobilenetv1 = 'mobilenetv1_int8_model.tar.gz'
    self.small_data = 'imagenet_val_100_tail.tar.gz'
    self.url_base = 'http://paddle-inference-dist.bj.bcebos.com/int8/'
    urls = [
        self.url_base + 'imagenet_val_100_tail.tar.gz',
        self.url_base + 'mobilenetv1_int8_model.tar.gz'
    ]
    md5s = [
        'b6e05365252f12f75e7daf3fcbc62e96',
        '13892b0716d26443a8cdea15b3c6438b'
    ]
    for url, md5 in zip(urls, md5s):
        download(url, self.cache_folder, md5)
        file_name = url.split('/')[-1]
        tar_file = tarfile.open(os.path.join(self.cache_folder, file_name))
        target_dir = file_name.split('.')[0]
        target_dir = os.path.join(self.cache_folder, target_dir)
        tar_file.extractall(target_dir)
``` 
CI errors
``` txt
[03:14:52] :	 [Step 1/1]     r = requests.get(url, stream=True)
[03:14:52] :	 [Step 1/1]   File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/requests/api.py", line 71, in get
[03:14:52] :	 [Step 1/1]     return request('get', url, params=params, **kwargs)
[03:14:52] :	 [Step 1/1]   File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/requests/api.py", line 57, in request
[03:14:52] :	 [Step 1/1]     return session.request(method=method, url=url, **kwargs)
[03:14:52] :	 [Step 1/1]   File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/requests/sessions.py", line 475, in request
[03:14:52] :	 [Step 1/1]     resp = self.send(prep, **send_kwargs)
[03:14:52] :	 [Step 1/1]   File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/requests/sessions.py", line 585, in send
[03:14:52] :	 [Step 1/1]     r = adapter.send(request, **kwargs)
[03:14:52] :	 [Step 1/1]   File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/requests/adapters.py", line 442, in send
[03:14:52] :	 [Step 1/1]     raise ConnectionError(e, request=request)
[03:14:52] :	 [Step 1/1] requests.exceptions.ConnectionError: HTTPConnectionPool(host='paddle-inference-dist.bj.bcebos.com', port=80): Max retries exceeded with url: /int8/imagenet_val_100_tail.tar.gz (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x14258f438>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known',))
[03:14:52] :	 [Step 1/1] 
```

luotao1 commented:
We will look at the network problem of mac ci machines.

luotao1 commented on python/paddle/fluid/contrib/slim/tests/test_mkldnn_int8_quantization_strategy.py:
```diff
@@ -0,0 +1,127 @@
+from paddle.dataset.common import download
+
+
+class TestInferQuantizeStrategy(unittest.TestCase):
```
Why this UT enabled on MAC/windows? It should be on LInux only. 
Please see CMakelists.txt.

zhaify replied:
understand much thanks!

luotao1 commented on python/paddle/fluid/contrib/slim/tests/test_mkldnn_int8_quantization_strategy.py:
```diff
@@ -0,0 +1,127 @@
+    Test API of Post Training quantization strategy for int8 with MKLDNN.
+    """
+
+    def setUp(self):
```
duplicated codes in `test_calibration_resnet50.py`, please refine it.

zhaify replied:
done

luotao1 commented:
```
[03:38:23]	Cache file /Users/paddle/.cache/paddle/dataset/int8/download/imagenet_val_100_tail.tar.gz not found, downloading http://paddle-inference-dist.bj.bcebos.com/int8/imagenet_val_100_tail.tar.gzE
```
Besides, the download name is not right: `imagenet_val_100_tail.tar.gzE`.

zhaify commented:
@luotao1 CI passed, should we start to review this PR ?

sfraczek commented on paddle/fluid/inference/api/analysis_predictor_tester.cc:
```diff
@@ -260,7 +260,8 @@ class MkldnnQuantizerTest : public testing::Test {
     auto* predictor_p = static_cast<AnalysisPredictor*>(predictor.get());
 
-    auto qconfig = std::make_shared<MkldnnQuantizerConfig>();
+    // The MkldnnQuantizerConfig will be deleted by MkldnnQuantizer's destructor
```
Hey, this comment is no longer necessary. :)

zhaify replied:
done

zhaify commented:
@wojtuss  @sfraczek Can you also help to review this PR?

luotao1 commented on paddle/fluid/inference/api/analysis_config.cc:
```diff
@@ -168,11 +168,10 @@ void AnalysisConfig::EnableMkldnnQuantizer() {
   Update();
```
why do you change `analysis_config.cc`, `mkldnn_quantizer.h` and `paddle_analysis_config.h`?

zhaify replied:
these changes can avoid share_ptr double free.
@wojtuss @bingyanghuang

luotao1 commented on paddle/fluid/pybind/inference_api.cc:
```diff
@@ -249,13 +257,41 @@ void BindAnalysisConfig(py::module *m) {
       .def("to_native_config", &AnalysisConfig::ToNativeConfig)
+#ifdef PADDLE_WITH_MKLDNN
+      .def("enable_quantizer", &AnalysisConfig::EnableMkldnnQuantizer)
+      .def("quantizer_enabled", &AnalysisConfig::mkldnn_quantizer_enabled)
```
Do you need polish `quantizer_enabled`?
- it is similar to `enable_quantizer`
- I don't see its usage in this PR.

zhaify replied:
quantizer_enabled removed

luotao1 commented on python/paddle/fluid/contrib/slim/quantization/infer_quant_strategy.py:
```diff
@@ -0,0 +1,125 @@
+logger = logging.getLogger(__name__)
+
+
+class InferQuantStrategy(Strategy):
```
The name `InferQuantStrategy` and `infer_quant_strategy.py` are not better, since it means Post Training quantization strategy.

zhaify replied:
how about 
`MKLDNNPostTrainingQuantStrategy`
`mkldnn_post_training_strategy.py`

luotao1 replied:
I think it is OK. @bingyanghuang how do you think about?

bingyanghuang replied:
> I think it is OK. @bingyanghuang how do you think about?

Looks fine to me.

zhaify replied:
done

luotao1 commented on python/paddle/fluid/contrib/slim/quantization/infer_quant_strategy.py:
```diff
@@ -0,0 +1,125 @@
+                 fp32_model_path=None):
+        """
+        Args:
+            start_epoch(int): The 'on_epoch_begin' function will be called in start_epoch. default: 0
```
where is `on_epoch_begin` function?

luotao1 commented on python/paddle/fluid/contrib/slim/quantization/infer_quant_strategy.py:
```diff
@@ -0,0 +1,125 @@
+        """
+        Args:
+            start_epoch(int): The 'on_epoch_begin' function will be called in start_epoch. default: 0
+            end_epoch(int): The 'on_epoch_end' function will be called in end_epoch. default: 0
```
where is `on_epoch_end` funcition.

zhaify replied:
removed

luotao1 commented on python/paddle/fluid/contrib/slim/quantization/infer_quant_strategy.py:
```diff
@@ -0,0 +1,125 @@
+            start_epoch(int): The 'on_epoch_begin' function will be called in start_epoch. default: 0
+            end_epoch(int): The 'on_epoch_end' function will be called in end_epoch. default: 0
+            int8_model_save_path(str): The path to save model with int8_t weight.
+                            None means it doesn't save int8 model. defalut: None.
```
> The path to save model with int8_t weight.
None (?) means it doesn't save int8 model

- it's confusing. do you mean fp32 programdesc + int8 weight?
- int8_t ->int8

zhaify replied:
> * it's confusing. do you mean fp32 programdesc + int8 weight?

it will save int8 programdesc + fp32 weight, because mkldnn conv int8 kernel only support fp32 weight now.

luotao1 replied:
Got it. But the `int8_model_save_path` and its notes are too confusing. @bingyanghuang

zhaify replied:
I will refine the notes

luotao1 commented on python/paddle/fluid/contrib/slim/quantization/infer_quant_strategy.py:
```diff
@@ -0,0 +1,125 @@
+            image.name = "x"
+            image.shape = shape
+            image.dtype = core.PaddleDType.FLOAT32
+            image.data.resize(num_images * 3 * 224 * 224 * sys.getsizeof(float))
```
`[3, 224, 224]` should not be in a common function.

zhaify replied:
done

zhaify replied:
done

luotao1 commented on python/paddle/fluid/contrib/slim/tests/test_mkldnn_int8_quantization_strategy.py:
```diff
@@ -0,0 +1,112 @@
+        com_pass = Compressor(
+            None,
+            None,
+            None,
```
please write line93-95 as line96, we don't know the parameter name of None

luotao1 commented on python/paddle/fluid/contrib/slim/tests/test_mkldnn_int8_quantization_strategy.py:
```diff
@@ -0,0 +1,112 @@
+        with open(config_path, 'r+') as fp:
+            data = fp.read()
+        model_path = os.path.join(self.cache_folder,
+                                  'mobilenetv1_int8_model/model')
```
please add resnet50 test as well.

zhaify replied:
done

luotao1 commented on python/paddle/fluid/contrib/slim/tests/test_mkldnn_int8_quantization_strategy.py:
```diff
@@ -0,0 +1,112 @@
+                    yield img, int(label)
+                    step += 1
+                    if cycle and step == num:
+                        step = 0
```
could you reuse the reader in `test_calibraion_resnet50.py`?

zhaify replied:
if reuse, the result (top1, top5) will not align with CAPI test (analyzer_int8_image_classification_tester.cc) since CAPI test use "data.bin" file for it's test.

luotao1 replied:
Do you mean only the dataset different?
https://github.com/PaddlePaddle/Paddle/blob/e4a52e082fe84a4713d05370f077eeda413a870c/python/paddle/fluid/contrib/tests/test_calibration_resnet50.py#L85-L90
I see the parameter has `file_list` and `data_dir` to pass the dataset.

zhaify replied:
> Do you mean only the dataset different? **Right**

> I see the parameter has `file_list` and `data_dir` to pass the dataset.

`file_list` parameter need a file contains all the files name like
``` txt
val/ILSVRC2012_val_00000093.JPEG 366
val/ILSVRC2012_val_00000094.JPEG 780
...
```
`data_dir` is the root path to all the pictures list in `file_list`,
the reader in test_calibration_resnet50.py will read each picture by path (data_dir + val/ILSVRC2012_val_*.JPEG), so i think it cannot pass "data.bin"

luotao1 commented:
please use `start a review` to reply comments
http://paddlepaddle.org/documentation/docs/zh/1.4/advanced_usage/development/contribute_to_paddle/submit_pr_guide.html#permalink-8--
![image](https://user-images.githubusercontent.com/6836917/58456517-08208e80-8157-11e9-9c39-ce9fc20172b8.png)

wojtuss commented on paddle/fluid/pybind/inference_api.cc:
```diff
@@ -249,13 +257,40 @@ void BindAnalysisConfig(py::module *m) {
            &AnalysisConfig::cpu_math_library_num_threads)
       .def("to_native_config", &AnalysisConfig::ToNativeConfig)
+#ifdef PADDLE_WITH_MKLDNN
+      .def("enable_quantizer", &AnalysisConfig::EnableMkldnnQuantizer)
```
The `EnableMkldnnQuantizer()` method is defined also without MKL-DNN, so I think we do not have to wrap it with the `#ifdef`.

zhaify replied:
done

wojtuss commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,108 @@
+# Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
`2019`

zhaify replied:
done

wojtuss commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,108 @@
+
+    def on_compression_begin(self, context):
+        """
+	Prepare the data and quantify the model
```
Please add indentation to the comment

zhaify replied:
done

wojtuss replied:
It is still not aligned with `"""`.

zhaify replied:
done

wojtuss commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,108 @@
+        """
+        super(MKLDNNPostTrainingQuantStrategy,
+              self).on_compression_begin(context)
+        logger.info('InferQuantStrategy::on_compression_begin')
```
`InferQuantStrategy` name is not used anymore. `MKLDNNPostTrainingQuantStrategy` should be used instead?

zhaify replied:
done

wojtuss commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,108 @@
+        infer_config.quantizer_config().set_quant_batch_size(num_images)
+        #Run INT8 MKL-DNN Quantization
+        predictor = core.create_paddle_predictor(infer_config)
+        if self.int8_model_save_path and self.int8_model_save_path != '':
```
empty string gives `false`, the first check should be enough

zhaify replied:
done

wojtuss commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,108 @@
+                os.makedirs(self.int8_model_save_path)
+            predictor.SaveOptimModel(self.int8_model_save_path)
+
+        logger.info('Finish InferQuantStrategy::on_compresseion_begin')
```
`InferQuantStrategy` name is not used anymore

zhaify replied:
done

wojtuss commented on python/paddle/fluid/contrib/slim/tests/test_mkldnn_int8_quantization_strategy.py:
```diff
@@ -0,0 +1,134 @@
+            train_optimizer=None,
+            distiller_optimizer=None)
+        com_pass.config(config_path)
+        eval_graph = com_pass.run()
```
Please add the accuracy result after the quantization. Otherwise, we cannot verify the quantization was done properly. I know the INT8v2 quantization is verified by the c-api unit tests, but the check should be added here as well, in my opinion.

zhaify replied:
done

wojtuss commented on python/paddle/fluid/contrib/slim/tests/test_mkldnn_int8_quantization_strategy.py:
```diff
@@ -0,0 +1,134 @@
+        if not fluid.core.is_compiled_with_mkldnn():
+            return
+
+        base_url = 'http://paddle-inference-dist.bj.bcebos.com/int8/'
```
If it is possible, it would be good to reuse models and datasets downloaded for the c-api unit tests. They seem to be the same.
You can take a look at `paddle/fluid/inference/tests/api/CMakeLists.txt` and `paddle/fluid/inference/tests/test.cmake`.

zhaify replied:
done

wojtuss commented on python/paddle/fluid/contrib/slim/tests/CMakeLists.txt:
```diff
@@ -1,15 +1,57 @@
+
+  #resnet50 int8
+  set(INT8_RESNET50_MODEL_DIR "${INT8_DATA_DIR}/resnet50")
+  inference_analysis_python_api_int8_test(test_analyzer_int8_resnet50 ${INT8_RESNET50_MODEL_DIR} ${INT8_DATA_DIR} test_mkldnn_int8_quantization_strategy.py)
```
With this configuration running
```
ctest -R test_analyzer_int8_resnet50
```
executes both: c-api unit test and the python slim test for ResNet50.
Please, make those tests separate ones, i.e. rename the tests here `test_analyzer_int8_resnet50`->`test_slim_int8_resnet50`, and so on.

zhaify replied:
done

wojtuss commented on python/paddle/fluid/contrib/slim/tests/CMakeLists.txt:
```diff
@@ -1,15 +1,57 @@
+  inference_analysis_python_api_int8_test(test_analyzer_int8_googlenet ${INT8_GOOGLENET_MODEL_DIR} ${INT8_DATA_DIR} test_mkldnn_int8_quantization_strategy.py)
 endif()
 
+list(REMOVE_ITEM TEST_OPS test_mkldnn_int8_quantization_strategy)
```
Please, add a comment on why the test is disabled.

zhaify replied:
done

luotao1 replied:
I still don't understand line53: 
- why remove this test here? 
- Do you need to move this test after line 22?
- current CI not support for testing, which machine could test it?

wojtuss commented on python/paddle/fluid/contrib/slim/tests/test_mkldnn_int8_quantization_strategy.py:
```diff
@@ -93,42 +153,59 @@ def _test_int8_quant(self, data_path, config_path):
+            self._reader_creator(data_path, False), batch_size=batch_size)
+        int8_model_result = self._predict(val_reader, int8_model_path)
+        _logger.info('--- comparing outputs ---')
+        _logger.info('Avg top1 INT8 accuracy: {0:.4}'.format(int8_model_result[
```
Please, make it printing accuracy with at least 4 digits precision after period, padding with zeros if necessary.

zhaify replied:
done

wojtuss commented on python/paddle/fluid/contrib/slim/tests/test_mkldnn_int8_quantization_strategy.py:
```diff
@@ -12,22 +12,52 @@
+        default=100,
+        help='batch size for quantization warmup')
+    parser.add_argument(
+        '--quantized_accuracy',
```
Please make the option name more meaningful, e.g. `--accuracy_diff_threshold`.

zhaify replied:
done

wojtuss commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,104 @@
+        labels.data = core.PaddleBuf(label_data)
+
+        warmup_data = [images, labels]
+
```
Please, make it running on multiple cores by adding
```
infer_config.set_cpu_math_library_num_threads(???)
```
with a proper argument from `FLAGS_OMP_NUM_THREADS` env variable. Otherwise, the test does not utilize multiple cores.

zhaify replied:
done

wojtuss commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,104 @@
+
+    def on_compression_begin(self, context):
+        """
+	    Prepare the data and quantify the model
```
Please, make the comment aligned with indentation of `"""`.

wojtuss commented on python/paddle/fluid/contrib/slim/tests/test_mkldnn_int8_quantization_strategy.py:
```diff
@@ -0,0 +1,211 @@
+            0]))
+        _logger.info('Accepted accuracy drop threshold: {0}'.format(
+            quantized_accuracy))
+        assert abs(int8_model_result[0] - fp32_model_result[
```
Please, modify the assert condition so it pass only when
```
fp32_model_result[0] - int8_model_result[0] <= quantized_accuracy
```
It may happen that int8 accuracy is higher than fp32 accuracy, and that is OK.

zhaify replied:
done

luotao1 commented on python/paddle/fluid/contrib/slim/tests/quantization/config_mkldnn_int8.yaml:
```diff
@@ -0,0 +1,11 @@
+    mkldnn_post_training_strategy:
+        class: 'MKLDNNPostTrainingQuantStrategy'
+        int8_model_save_path: 'OUTPUT_PATH'
+        fp32_model_path: 'MODEL_PATH'
```
Do you need `cpu_math_library_num_threads` here, i,e do not directly use `FLAGS_omp_num_thread` to change it? @bingyanghuang @wojtuss 
Maybe you can update in next PR.

wojtuss replied:
@luotao1 , I think you are right, the number of threads could be set here instead of using the `FLAGS_OMP_NUM_THREADS` environment variable.

zhaify replied:
done

luotao1 commented:
```
Test #651: test_slim_int8_mobilenet .....................   Passed   62.47 sec
Test #653: test_slim_int8_resnet101 .....................   Passed  162.41 sec
Test #656: test_slim_int8_googlenet ......................   Passed   63.16 sec
Test #652: test_slim_int8_mobilenetv2 ....................   Passed   78.97 sec
Test #650: test_slim_int8_resnet50 .........................   Passed   78.49 sec
Test #655: test_slim_int8_vgg19 .........................   Passed  234.20 sec
Test #654: test_slim_int8_vgg16 ............................   Passed  153.20 sec
```
These tests cost a lot of time on CI, do you have any solutions to speedup CI time?

bingyanghuang commented:
> ```
> Test #651: test_slim_int8_mobilenet .....................   Passed   62.47 sec
> Test #653: test_slim_int8_resnet101 .....................   Passed  162.41 sec
> Test #656: test_slim_int8_googlenet ......................   Passed   63.16 sec
> Test #652: test_slim_int8_mobilenetv2 ....................   Passed   78.97 sec
> Test #650: test_slim_int8_resnet50 .........................   Passed   78.49 sec
> Test #655: test_slim_int8_vgg19 .........................   Passed  234.20 sec
> Test #654: test_slim_int8_vgg16 ............................   Passed  153.20 sec
> ```
> 
> These tests cost a lot of time on CI, do you have any solutions to speedup CI time?

I have some suggestions: 1. use more omp num threads; 2. Decrease the accuracy test batch size, need test the influence for the accuracy comparison @zhaify 3. @wojtuss  could we decrease the warmup batch size < 100?

luotao1 commented:
> use more omp num threads

We could not use more than 4 threads. Since the least cpu cores on ci machine are 16, and ci parallel level is 4.

luotao1 commented:
@bingyanghuang @wojtuss Could we reduce the unit-tests and only choose the least one or two time-cost unit-tests such as mobilenet and googlenet? The reason is:
- The core computation of these tests are the same in the `analyzer_test_xxx`
- This PR is only for python wrapper, don't focus on the int8 computation.

wojtuss commented:
> I have some suggestions: 1. use more omp num threads; 2. Decrease the accuracy test batch size, need test the influence for the accuracy comparison @zhaify 3. @wojtuss could we decrease the warmup batch size < 100?
@bingyanghuang , @luotao1 , usually warmup batch size 50 or even less is enough for the test to pass. It has to be verified per model.

qingqing01 commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,109 @@
+        labels.shape = [num_images, 1]
+        labels.dtype = core.PaddleDType.INT64
+        label_data = [label for (_, label) in data]
+        labels.data = core.PaddleBuf(label_data)
```
1. Here, the image and label will limit  the `MKLDNNPostTrainingQuantStrategy` only can be used for image classification.
2. Maybe need add `TODO` to remove this limits?

bingyanghuang replied:
Thanks qingqing for your review. We will add `TODO` to remove this limits.  @zhaify Please follows qiangqiang' suggestions.

zhaify replied:
done

bingyanghuang commented:
> @bingyanghuang @wojtuss Could we reduce the unit-tests and only choose the least one or two time-cost unit-tests such as mobilenet and googlenet? The reason is:
> 
> * The core computation of these tests are the same in the `analyzer_test_xxx`
> * This PR is only for python wrapper, don't focus on the int8 computation.

The reason why we add these unit test in the python is to make sure that there is no diff for the results between C-API and Python API. @luotao1  If you think that one or two models tests can be enough for this results alignment test, we can reduce this test to 1 or 2 the least time-consuming test at once.

wzzju commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,109 @@
+    def __init__(self, int8_model_save_path=None, fp32_model_path=None):
+        """
+        Args:
+            int8_model_save_path(str): The path to save int8 model with fp32 weights.
```
I don't understand what is the `int8 model with fp32 weights`?Please give more detail explanation.

luotao1 replied:
The same to me. I comment in https://github.com/PaddlePaddle/Paddle/pull/17634#discussion_r287941929, but it seems still confusing here.

zhaify replied:
It means `MKLDNNPostTrainingQuantStrategy` will save **int8 programdesc + fp32 weights** to the path `int8_model_save_path`

bingyanghuang replied:
> It means `MKLDNNPostTrainingQuantStrategy` will save **int8 programdesc + fp32 weights** to the path `int8_model_save_path`

@zhaify we should make the instructions/explanation in code more clearly.

zhaify replied:
done

wzzju commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,109 @@
+        Args:
+            int8_model_save_path(str): The path to save int8 model with fp32 weights.
+                            None means it doesn't save int8 model. defalut: None.
+            fp32_model_path(str): The path to model with fp32 weight. defalut: None.
```
The same above.

zhaify replied:
For post training quantization, MKLDNNPostTrainingQuantStrategy support to convert a **fp32 programdesc with fp32 weights** to a **int8 programdesc with fp32 weights**, so fp32_model_path means the path to original fp32 model need to be quantized.

bingyanghuang replied:
> The same above.

Will refactor the args instructions for them.

zhaify replied:
done

luotao1 replied:
Do you refactor the args instructions? I see it is the same as before. The comments https://github.com/PaddlePaddle/Paddle/pull/17634#discussion_r289680391 and https://github.com/PaddlePaddle/Paddle/pull/17634#discussion_r289679585 are more clear than the args notes.

zhaify replied:
done

luotao1 commented:
@bingyanghuang I think you can reduce the ci time by
- first try https://github.com/PaddlePaddle/Paddle/pull/17634#issuecomment-498071688?
- if above try fails, then reduce this test to 1 or 2 the least time-consuming test at once. If you need 
 to reduce the tests, you could use some flags to make our QA tests them in local machine.

bingyanghuang commented:
> @bingyanghuang I think you can reduce the ci time by
> 
> * first try [#17634 (comment)](https://github.com/PaddlePaddle/Paddle/pull/17634#issuecomment-498071688)?
> * if above try fails, then reduce this test to 1 or 2 the least time-consuming test at once. If you need
>   to reduce the tests, you could use some flags to make o.ur QA tests them in local machine.

Got it. Thanks

wzzju commented on python/paddle/fluid/contrib/slim/tests/quantization/mkldnn_int8_quantization_strategy.py:
```diff
@@ -0,0 +1,212 @@
+#   copyright (c) 2019 paddlepaddle authors. all rights reserved.
```
Is this file a unit test? If so, it should be placed in the `python/paddle/fluid/contrib/slim/tests/` path. What's more, its name should be started with `test_`. Otherwise, it will not be run in the `CI`.

wzzju replied:
`python/paddle/fluid/contrib/slim/tests/quantization` should only place some yaml config files, not python files.

zhaify replied:
This is the suggestion comes form @wojtuss, he think that adding `test_mkldnn_int8_quantization_strategy.py` in `python/paddle/fluid/contrib/slim/tests/` and removed if not running on Linux platform compiled with MKLDNN is a bad idea.

wzzju commented on python/paddle/fluid/contrib/slim/tests/quantization/mkldnn_int8_quantization_strategy.py:
```diff
@@ -0,0 +1,212 @@
+        return reader
+
+    def _update_config_file(self, fp32_model_path, output_path):
+        config_path = './quantization/config_mkldnn_int8.yaml'
```
Do you run this unit test file locally? If this file in `quantization` directory,  I don't think this UT can be run successfully because of the `config_path` value. When you raise a pull request, please be sure that its UT can be run successfully in your local machine.

zhaify replied:
It runs ok with cmd `ctest -R test_slim_int8_*` on my local machine.

zhaify commented on python/paddle/fluid/contrib/slim/tests/CMakeLists.txt:
```diff
@@ -1,15 +1,57 @@
+  inference_analysis_python_api_int8_test(test_analyzer_int8_googlenet ${INT8_GOOGLENET_MODEL_DIR} ${INT8_DATA_DIR} test_mkldnn_int8_quantization_strategy.py)
 endif()
 
+list(REMOVE_ITEM TEST_OPS test_mkldnn_int8_quantization_strategy)
```
@wzzju how about this version of cmake

zhaify replied:
done

luotao1 replied:
how about https://github.com/PaddlePaddle/Paddle/pull/17634#discussion_r290310558?

zhaify replied:
> I still don't understand line53:

I have write a comment in CMakeList.txt before and refactor again. 
> * why remove this test here?


> * Do you need to move this test after line 22?

test_mkldnn_int8_quantization_strategy only support for testing on Linux platform with MKL-DNN enabled, otherwise it should be disabled (here is removed). If on Linux and enable MKL-DNN test_mkldnn_int8_quantization_strategy has been tested from line 20 -line 51.

> * current CI not support for testing, which machine could test it?

current CI means: on MAC OS is MAC_PR_CI, On windows OS is PR_Windows_CI.

luotao1 commented on python/paddle/fluid/contrib/slim/tests/quantization/config_mkldnn_int8.yaml:
```diff
@@ -4,6 +4,7 @@ strategies:
         class: 'MKLDNNPostTrainingQuantStrategy'
         int8_model_save_path: 'OUTPUT_PATH'
         fp32_model_path: 'MODEL_PATH'
+        omp_num_threads: 4
```
How about `cpu_math_library_num_threads`?

zhaify replied:
sounds good, I will change the name.

zhaify replied:
done

luotao1 commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,116 @@
+                 cpu_math_library_num_threads=1):
+        """
+        Args:
+            int8_model_save_path(str): The path to save model for MKL-DNN int8 inference.
```
I see there is no refactor.

zhaify replied:
done with your [suggestion](https://github.com/PaddlePaddle/Paddle/pull/17634#discussion_r290305425)

luotao1 commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,116 @@
+                            None means it doesn't save model. defalut: None.
+            fp32_model_path(str): The path of fp32 model to be converted to int8 model
+                            None means it doesn't have fp32 model. defalut: None.
+            cpu_math_library_num_threads(int): The num of omp threads user want 
```
The number of cpu math library threads.

zhaify replied:
done

luotao1 commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,116 @@
+            fp32_model_path(str): The path of fp32 model to be converted to int8 model
+                            None means it doesn't have fp32 model. defalut: None.
+            cpu_math_library_num_threads(int): The num of omp threads user want 
+                            MKLDNNPostTrainingQuantStrategy to use. defalut: 1.
```
defalut->default. @zhaify @bingyanghuang Please check all the notes again, thanks very much!

zhaify replied:
done

luotao1 commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,116 @@
+                            None means it doesn't have fp32 model. defalut: None.
+            cpu_math_library_num_threads(int): The num of omp threads user want 
+                            MKLDNNPostTrainingQuantStrategy to use. defalut: 1.
+                            1 means only use one omp thread.
```
1 means only use one omp thread.
There are two verbs （means and use）in one sentence.

zhaify replied:
done

luotao1 commented on python/paddle/fluid/contrib/slim/tests/CMakeLists.txt:
```diff
@@ -1,11 +1,59 @@
+        ARGS --infer_model ${model_dir}/model
+             --infer_data ${data_dir}/data.bin
+             --int8_model_save_path int8_models/${target}
+             --warmup_batch_size 100
```
I see you change it to 10 in https://github.com/PaddlePaddle/Paddle/pull/17634/commits/085642a9fdba219180e2084a8ef5bf2d117b207f#diff-1d840481eb99c8cb8ae2116875d02d57R9. Why do you change back to 100 again?

zhaify replied:
I have tried  `warmup_batch_size 10, 20,30,...100`, and `warmup_batch_size=10` is a very special  case that can pass CI (all the 7 models) which can reduce time cost from 30s+ to several seconds (googlenet) on our local machine, @bingyanghuang think it is not a good idea, so I change it back and reduce the number of images used in accuracy check from 100 to 10 that also can reduce some time cost by CI. 
And if you prefer the previous solution, I can rollback it.

luotao1 replied:
Got it. But are 10 images too less to ensure the accuracy?

zhaify replied:
done with remaining googlenet & mobilenetv1 UTs and remove others,
and add cmake flags `WITH_SLIM_MKLDNN_FULL_TEST` for QA testing all models locally!

luotao1 commented on python/paddle/fluid/contrib/slim/tests/quantization/config_mkldnn_int8.yaml:
```diff
@@ -0,0 +1,12 @@
+        class: 'MKLDNNPostTrainingQuantStrategy'
+        int8_model_save_path: 'OUTPUT_PATH'
+        fp32_model_path: 'MODEL_PATH'
+        cpu_math_library_num_threads: 4
```
please comment why do you use 4 here.

zhaify replied:
done

luotao1 commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,120 @@
+                 cpu_math_library_num_threads=1):
+        """
+        Args:
+            int8_model_save_path(str): The path to save an int8 programdesc with fp32 weights for
```
1. programdesc-> ProgramDesc , please correct all.
2. `The path to save xxx`, is grammar right？

zhaify replied:
done

luotao1 commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,120 @@
+            int8_model_save_path(str): The path to save an int8 programdesc with fp32 weights for
+                        MKL-DNN int8 inference. For post training quantization,
+                        MKLDNNPostTrainingQuantStrategy only support to convert a fp32 programdesc
+                        with fp32 weights to a int8 programdesc with fp32 weights now. The saved
```
a int8 -> an int8

zhaify replied:
done

luotao1 commented on python/paddle/fluid/contrib/slim/quantization/mkldnn_post_training_strategy.py:
```diff
@@ -0,0 +1,120 @@
+                        None means it doesn't save model. default: None.
+            fp32_model_path(str): The path to the original fp32 programdesc with fp32 weights.
+                        None means it doesn't have a fp32 programdesc with fp32 weights. default: None.
+            cpu_math_library_num_threads(int): The number of cpu math library threads are used on 
```
which are used on

zhaify replied:
done

wojtuss commented on python/paddle/fluid/contrib/slim/tests/test_mkldnn_int8_quantization_strategy.py:
```diff
@@ -0,0 +1,220 @@
+    def _reader_creator(self,
+                        data_file='data.bin',
+                        cycle=False,
+                        predict_num=100):
```
Please, add an option to set the number of images to process. Currently, the value of 10 is hardcoded in lines 199 and 203, and cannot be modified.

zhaify replied:
done with removed, since reduce predict number of images in accuracy check is not a good idea.

wojtuss replied:
OK. This can be done later.

sfraczek commented on python/paddle/fluid/contrib/slim/tests/CMakeLists.txt:
```diff
@@ -1,11 +1,63 @@
 
+function(inference_analysis_python_api_int8_test target model_dir data_dir filename)
+    py_test(${target} SRCS ${filename}
+        ENVS FLAGS_OMP_NUM_THREADS=${CPU_NUM_THREADS_ON_CI}
```
Unfortunately it always runs on 1 thread regardless of  this parameter :(

sfraczek replied:
ok this can be added later

luotao1 commented:
@zhaify @bingyanghuang Please do refactor in next PR.
PR1: 
1. https://github.com/PaddlePaddle/Paddle/pull/17634#pullrequestreview-246201587
2. Investigate the CI unit test running time problem

PR2: 

3. Finish the test doc with performance and accuracy data
